pkg/webhook/admission/validator_test.go

```diff
@@ -0,0 +1,280 @@
+
+		})
+
+		It("should return 200 in response when delete succeeds", func() {
+
+			handler := createSucceedingValidatingHandler()
+
+			response := handler.Handle(context.TODO(), Request{
+				AdmissionRequest: v1beta1.AdmissionRequest{
+					Operation: v1beta1.Delete,
+					OldObject: runtime.RawExtension{
+						Raw:    []byte("{}"),
+						Object: handler.validator,
+					},
+				},
+			})
+			Expect(response.Allowed).Should(BeTrue())
+			Expect(response.Result.Code).Should(Equal(int32(http.StatusOK)))
+		})
+
+		It("should return 400 in response when delete fails on decode", func() {
+			//TODO
+		})
+
+		It("should return response built with the Status object when ValidateDelete returns APIStatus error", func() {
+
+			handler, expectedError := createValidatingHandlerWhichReturnsStatusError()
+
+			response := handler.Handle(context.TODO(), Request{
+				AdmissionRequest: v1beta1.AdmissionRequest{
+					Operation: v1beta1.Delete,
+					OldObject: runtime.RawExtension{
+						Raw:    []byte("{}"),
+						Object: handler.validator,
+					},
+				},
+			})
+			Expect(response.Allowed).Should(BeFalse())
+
+			apiStatus, ok := expectedError.(apierrs.APIStatus)
+			Expect(ok).Should(BeTrue())
+			Expect(response.Result.Code).Should(Equal(apiStatus.Status().Code))
+			Expect(*response.Result).Should(Equal(apiStatus.Status()))
+
+		})
+
+		It("should return 403 response when ValidateDelete returns non-APIStatus error", func() {
+
+			handler, expectedError := createValidatingHandlerWhichReturnsRegularError()
+
+			response := handler.Handle(context.TODO(), Request{
+				AdmissionRequest: v1beta1.AdmissionRequest{
+					Operation: v1beta1.Delete,
+					OldObject: runtime.RawExtension{
+						Raw:    []byte("{}"),
+						Object: handler.validator,
+					},
+				},
+			})
+			Expect(response.Allowed).Should(BeFalse())
+			Expect(response.Result.Code).Should(Equal(int32(http.StatusForbidden)))
+			Expect(string(response.Result.Reason)).Should(Equal(expectedError.Error()))
+
+		})
+	})
+})
+
+type fakeValidator struct {
+	ErrorToReturn error `json:"ErrorToReturn,omitempty"`
+}
+
+var _ Validator = &fakeValidator{}
+
+var fakeValidatorVK = schema.GroupVersionKind{Group: "foo.test.org", Version: "v1", Kind: "fakeValidator"}
+
+func (v *fakeValidator) ValidateCreate() error {
+	return v.ErrorToReturn
+}
+
+func (v *fakeValidator) ValidateUpdate(old runtime.Object) error {
+	return v.ErrorToReturn
+}
+
+func (v *fakeValidator) ValidateDelete() error {
+	return v.ErrorToReturn
+}
+
+func (v *fakeValidator) GetObjectKind() schema.ObjectKind { return v }
+
+func (v *fakeValidator) DeepCopyObject() runtime.Object {
+	return &fakeValidator{ErrorToReturn: v.ErrorToReturn}
+}
+
+func (v *fakeValidator) GroupVersionKind() schema.GroupVersionKind {
+	return fakeValidatorVK
+}
+
+func (v *fakeValidator) SetGroupVersionKind(gvk schema.GroupVersionKind) {}
+
+func createSucceedingValidatingHandler() *validatingHandler {
+	decoder, _ := NewDecoder(scheme.Scheme)
```

Comment:
if you just make decoder a global, you can inline much of this, which might make improve readability.

Alternatively, something like this:

```go
Describe("validating handler", func() {
   Context("when dealing with status errors", func() {
       handler := &fakeHandler{...}
       It("should propagate the status from ValidateDelete's return value to the HTTP response", func() { ... })
       // etc
   })
   Context("when dealing with non-status errors", func() { ... })
   Context("when dealing with successful results", func() { ... })
})
```

Reply:
Fixed